crates/pop-wallet-integration/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+use axum::{
+	routing::{get, post},
+	Router,
+};
+use serde::Serialize;
+use std::{path::PathBuf, sync::Arc};
+use tokio::sync::{oneshot, Mutex};
+use tower_http::services::ServeDir;
+
+/// Data to be sent to frontend for signing.
+#[derive(Serialize, Debug)]
+pub struct Data {
+	chain_rpc: String,
+	data_type: DataType,
+}
+
+/// The type of transaction with specific data for signing (contract or parachain).
+#[derive(Serialize, Debug, PartialEq)]
+pub enum DataType {
+	/// Parachain call, where Vec<u8> is the encoded call data.
+	Parachain(Vec<u8>),
+	Contract(ContractArgs),
+}
+
+/// Contract specific data variations.
+pub mod contracts {
+	use super::Serialize;
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct ContractArgs {
+		call_type: ContractCallType,
+		storage_deposit_limit: Option<String>,
+		/// The binary (wasm, polkavm) of the contract.
+		code: Option<Vec<u8>>,
+	}
+
+	#[derive(Serialize, Debug, PartialEq)]
+	pub enum ContractCallType {
+		// no unique fields.
+		Upload,
+		Instantiate(InstantiateArgs),
+		Call(CallArgs),
+	}
+
+	/// Arguments for instantiating a contract.
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct InstantiateArgs {
+		constructor: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+		salt: Option<Vec<u8>>,
+	}
+
+	/// Arguments for calling a contract.
+	#[derive(Serialize, Debug, PartialEq)]
+	pub struct CallArgs {
+		address: String,
+		message: String,
+		args: Vec<String>,
+		value: String,
+		gas_limit: Option<u64>,
+		proof_size: Option<u64>,
+	}
+
+	#[cfg(test)]
+	mod tests {
+		use super::*;
+		use crate::*;
+		use serde_json::json;
+
+		#[test]
+		fn json_serialize_parachain_data_works() {
+			let data = Data {
+				chain_rpc: "localhost:9944".to_string(),
+				data_type: DataType::Parachain(vec![0, 1, 2, 3]),
+			};
+			let json_data = serde_json::to_value(&data).unwrap();
+
+			assert_eq!(
+				json_data,
+				json!({"chain_rpc": "localhost:9944", "data_type": {"Parachain": [0,1,2,3]}})
+			);
+		}
+		#[test]
+		fn json_serialize_contract_upload_data_works() {
+			let data = Data {
+				chain_rpc: "localhost:9944".to_string(),
+				data_type: DataType::Contract(ContractArgs {
+					call_type: ContractCallType::Upload,
+					storage_deposit_limit: Some("1234".to_string()),
+					code: Some(vec![0, 1, 2]),
+				}),
+			};
+			let json_data = serde_json::to_value(&data).unwrap();
+
+			assert_eq!(
+				json_data,
+				json!({"chain_rpc": "localhost:9944", "data_type": {"Contract": {
+					"call_type": "Upload",
+					"storage_deposit_limit": "1234",
+					"code": [0,1,2]
+				}}})
+			);
+		}
+
+		// TODO: after high level review, complete serialization tests.
+	}
+}
+use crate::contracts::ContractArgs;
+
+struct StateHandler {
```

Comment:
Comments would be appreciated to better understand the usages. My understanding is that a message is sent to the channel to shutdown the server once successfully signed.

Reply:
But if its being submitted by the CLI via subxt, then as long as its signed does it matter? Guess it depends on whether is the UI showing inclusion/finalization or the CLI. Or both?!

Reply:
Preference is Pop CLI only.